tests/regression/src/lib.rs

```diff
@@ -0,0 +1,168 @@
+            assert!(pair.handshake().is_ok());
+            ctrl.stop_instrumentation();
+            Ok(())
+        })
+        .unwrap();
+    }
+    // function to run specified test using valgrind
+    fn run_valgrind_test(test_name: &str) {
+        let exe_path = std::env::args().next().unwrap();
+        let _dir_path = create_dir(Path::new("target/cg_artifacts"));
+        let output_file = format!("target/cg_artifacts/cachegrind_{}.out", test_name);
+        let output_command = format!("--cachegrind-out-file={}", &output_file);
+        let mut command = Command::new("valgrind");
+        command
+            .args(["--tool=cachegrind", &output_command, &exe_path, test_name])
+            .env_remove("ENABLE_VALGRIND"); //ensures that the recursive call is made to the actual harness code block rather than back to this function
+
+        println!("Running command: {:?}", command);
+        let status = command.status().expect("Failed to execute valgrind");
+
+        if !status.success() {
+            panic!("Valgrind failed");
+        }
+
+        let annotate_output = Command::new("cg_annotate")
+            .arg(&output_file)
+            .output()
+            .expect("Failed to run cg_annotate");
+
+        if !annotate_output.status.success() {
+            panic!("cg_annotate failed");
+        }
+        let _dir_path = create_dir(Path::new("target/perf_outputs"));
+        let annotate_file = format!("target/perf_outputs/{}.annotated.txt", test_name);
+        let mut file = File::create(&annotate_file).expect("Failed to create annotation file");
+        file.write_all(&annotate_output.stdout)
+            .expect("Failed to write annotation file");
+
+        let count = grep_for_instructions(&annotate_file)
+            .expect("Failed to get instruction count from file");
+        //this is temporary code to showcase the future diff functionality, here the code regresses by 10% each time so this test will almost always fail
+        let new_count = count + count / 10;
+        let diff = new_count - count;
+        assert!(diff <= self::COST, "Instruction count difference in {} exceeds the threshold, regression of {} instructions", test_name, diff);
+    }
+
+    // parses the annotated file for the overall instruction count total
+    fn grep_for_instructions(file_path: &str) -> Result<u64, io::Error> {
+        let path = Path::new(file_path);
+        let file = File::open(path)?;
+        let reader = io::BufReader::new(file);
+
+        for line in reader.lines() {
+            let line = line?;
+            if line.contains("PROGRAM TOTALS") {
+                if let Some(instructions) = line.split_whitespace().next() {
+                    return instructions
+                        .replace(',', "")
+                        .parse::<u64>()
+                        .map_err(|e| io::Error::new(io::ErrorKind::InvalidData, e));
+                }
+            }
+        }
+
+        Err(io::Error::new(
```

Comment:
This seems like it would be an unrecoverable error, in which case I think we can just panic 🤷